Skipping pytree registration in case fsdp is enabled

romitjain opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #39795 

Pytree registration for Dynamic KV Cache prevents FSDP from sharding the Q, K, V cache matrices correctly. This leads to a memory explosion during training (see the attached issue).

This PR fixes it by moving the pytree registration of Dynamic KV Cache from `cache_utils` to `integrations.executorch`, where it is only executed in case `torch.export` is called. This is an improvement over the previous version, where this registration was happening globally.

As a side effect of this PR, I have also moved `is_fsdp_enabled` helper function to a more suitable location -> `integrations.fsdp` (from `modeling_utils`)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #39795 
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

@Cyrilvallez @SunMarc @gante

romitjain committed:
Skipping pytree registration in case fsdp is enabled

romitjain committed:
Skipping pytree registration in case fsdp is enabled

romitjain committed:
Beauty changes

romitjain committed:
Beauty changes

romitjain committed:
Beauty changes

romitjain committed:
Beauty changes

romitjain committed:
Merge branch 'main' into bugfix/fsdp-pytree-memory-regression

gante reviewed (approved):
LGTM (after fixing import issues, see above), thank you for fixing 🤗 

related: [my reply in the original issue](https://github.com/huggingface/transformers/issues/39795#issuecomment-3179793156), this is indeed the way to go about the original issue

romitjain committed:
Moved the is_fsdp_available function to import utils

romitjain committed:
Moved the is_fsdp_available function to import utils

romitjain committed:
Moved is_fsdp_available to integrations.fsdp

romitjain committed:
Moved is_fsdp_available to integrations.fsdp

romitjain committed:
Added pytree registration inside dynamic cache class

romitjain committed:
Making ci/cd lords happy

romitjain commented:
@gante @Cyrilvallez 

1. I have moved `is_fsdp_enabled` to integrations.fsdp to avoid circular import
2. I have moved PyTree registration inside the `DynamicCache` class since it was getting registered globally at the imports earlier. The first indication that the model is being run in FSDP is during its forward pass. During that time, it also creates the `DynamicCache` class (https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L371). Hence, this is the appropriate location to decide if we want to register pytees. It's controlled by class variables and methods so it won't lead to duplicate registrations.

Please have a look.

romitjain committed:
Adding a check if DynamicCache is already a leaf

romitjain committed:
Adding try/catch for multiple initializations of DynamicCache in test suites

romitjain committed:
Merge branch 'bugfix/fsdp-pytree-memory-regression' of github.com:romitjain/transformers into bugfix/fsdp-pytree-memory-regression

romitjain committed:
Merge branch 'bugfix/fsdp-pytree-memory-regression' of github.com:romitjain/transformers; branch 'main' of github.com:huggingface/transformers into bugfix/fsdp-pytree-memory-regression

romitjain committed:
Moving dynamic cache pytree registration to executorch

romitjain committed:
Adding try catch back

romitjain commented:
@gante This is ready for review now.

Essentially, I have moved the pytree registration to `integrations.executorch`, which will be called only when exporting with `DynamicCache`.

I avoided the global registration since `integrations.executorch` functions are being called in `src.transformers.__init__`

gante reviewed (approved):
Yup, that's it -- thank you for iterating 💛

gante commented:
@Cyrilvallez have a look at the PR too please 🤗

Cyrilvallez reviewed (approved):
LGTM! Thanks for the great cleanup @romitjain 🤗 Beyond fixing fsdp, it makes the pytree registration much clearer 🤗

SunMarc reviewed (commented):
Nice job !
